site-src/geps/gep-1282.md

```diff
@@ -98,12 +106,292 @@ Of course, from the section above, you can see that there are many other feature
+    // Valid values include:
+    // * "Service" (the default if no Kind is specified)
+    // * "ServiceImport"
+    //
+    // Invalid values include:
+    //
+    // * "invalid/kind" - "/" is an invalid character
+    // +required
+    Kind string `json:"kind"`
+
+    // Name is the name of the target resource.
+    Name ObjectName `json:"name"`
+}
+
+// BackendCapabilitiesSpec defines a group of specific properties held by its
+// affiliated backend, and specified by the backend owner (typically the 
+// application developer).  
+//
+// Examples of properties that belong in the BackendCapabilitiesSpec include:
+// TLS re-encryption, session state, timeouts, retries, Websockets, HTTP/2, 
+// load balancer algorithm, healthcheck info, external traffic policy,
+// connection limits, error page, protocol switch, alerts, and others. 
+//
+// Support: Extended
+//
+type BackendCapabilitiesSpec struct {
+    // Backend refers to the object to which we apply these capabilities
+	// +required
+	Backend backendTargetReference `json:"backend,omitempty"`
+	
+    // TLS is the section of details on encryption processing in the traffic 
+    // hop from Gateway to backend, sometimes known as upstream TLS or re-encrypt.
+    // A xRoute configured as passthrough cannot specify a TLS encrypt type.
+    // TLS is useful for HTTPRoute and should not be used on UDP/TCP/TLSRoute.
+    //
+    // +optional
+    TLS *EncryptType `json:"tls,omitempty"`
+}
+
+// EncryptType describes the details needed by the client (Gateway) to implement 
+// encryption from the client to the destination endpoint.
+type EncryptType struct {
+    // The minimal TLSVersion that is accepted by the endpoint.
+    // If empty, there is no minimal backend TLSVersion.
+    // +optional	
+    MinVersion   *TLSVersion `json:"minVersion,omitempty"`
+
+    // The maximum TLSVersion that is accepted by the endpoint.
+    // If empty, there is no maximum backend TLSVersion.
+    // +optional    
+    MaxVersion   *TLSVersion  `json:"maxVersion,omitempty"`
+    
+    // CipherSuites is the minimum cipher suite that the backend accepts.
+    // If empty, anything that is appropriate for the MinVersion is accepted.
+    // If the MinVersion is not specified, the result is implementation
+    // dependent, e.g. based on the minimimal TLS version supported by the
+    // implementation.
+    // +optional
+    CipherSuites []string `json:"cipherSuites,omitempty"`
+    
+    // CACerts is the PEM bundle containing Certificate Authority, trusted by both 
+    // client and endpoint, and intermediate certificates needed to complete the 
+    // certificate chain of trust.  If empty, the system root certs should be used.
+    //
+    // +optional	
+    CACerts      *SecretObjectReference `json:"caCerts,omitempty"`
+    
+    // ServiceNames provide multiple named domains which will share a certificate
+    // for the endpoint.
+    // +optional
+    ServiceNames []string `json:"serviceNames,omitempty"`
```

Comment:
Why ? How does it help the caller to know ahead of time the list that will be returned in the cert anyways ?

You mean 'SANNames' - for supporting spiffee or equivalent ? That may be useful. 
Do you mean this as an override of the normal TLS server verification - where gateway talk with foo.ns.svc.cluster.local but get a certificate for example.com - I think it's important to be very clear with this, security is tricky.

Reply:
In this case, it's to help the _Gateway implementation_ so that it can determine configuration validity before allowing any connections at all. I'm so-so on this one, because using it makes the hostname match a three-part one (Listener, HTTPRoute, BackendCapabilities), rather than a two-part one, which is already hard to do. The upside is that it means that implementations can easily check and make clear _at a configuration level_ when something is not going to work, rather than waiting to have client requests fail and needing to troubleshoot that.